# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/renderer.rs
# Commit: 669ddccd1c34
# Full Hash: 669ddccd1c3485afd6f492070a7e0f325f8cfe78
# Author: Dzmitry Malyshau <dmalyshau@mozilla.com>
# Date: 2020-05-09 09:14:23
# Description:
#   Bug 1631331 - Handle a case where WebRender can't initialize GPU cache r=jrmuizel
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D74462
# ==============================================================================

diff -r 3877f1af4fe4 -r 669ddccd1c34 gfx/wr/webrender/src/renderer.rs
--- a/gfx/wr/webrender/src/renderer.rs	Fri May 08 14:17:31 2020 +0000
+++ b/gfx/wr/webrender/src/renderer.rs	Fri May 08 19:57:36 2020 +0000
@@ -3430,7 +3430,10 @@
                 }
 
                 frame.profile_counters.reset_targets();
-                self.prepare_gpu_cache(frame);
+                if let Err(e) = self.prepare_gpu_cache(frame) {
+                    self.renderer_errors.push(e);
+                    continue;
+                }
                 assert!(frame.gpu_cache_frame_id <= self.gpu_cache_frame_id,
                     "Received frame depends on a later GPU cache epoch ({:?}) than one we received last via `UpdateGpuCache` ({:?})",
                     frame.gpu_cache_frame_id, self.gpu_cache_frame_id);
@@ -3710,11 +3713,11 @@
         counters.updated_blocks.set(updated_blocks);
     }
 
-    fn prepare_gpu_cache(&mut self, frame: &Frame) {
+    fn prepare_gpu_cache(&mut self, frame: &Frame) -> Result<(), RendererError> {
         if self.pending_gpu_cache_clear {
             let use_scatter =
                 matches!(self.gpu_cache_texture.bus, GpuCacheBus::Scatter { .. });
-            let new_cache = GpuCacheTexture::new(&mut self.device, use_scatter).unwrap();
+            let new_cache = GpuCacheTexture::new(&mut self.device, use_scatter)?;
             let old_cache = mem::replace(&mut self.gpu_cache_texture, new_cache);
             old_cache.deinit(&mut self.device);
             self.pending_gpu_cache_clear = false;
@@ -3732,6 +3735,8 @@
             self.gpu_cache_texture.texture.as_ref().unwrap(),
             Swizzle::default(),
         );
+
+        Ok(())
     }
 
     fn update_texture_cache(&mut self) {
